internal/auth/internal/types/traffic_permissions.go

```diff
@@ -31,114 +32,206 @@ func RegisterTrafficPermissions(r resource.Registry) {
 		Proto:    &pbauth.TrafficPermissions{},
 		Scope:    resource.ScopeNamespace,
 		Validate: ValidateTrafficPermissions,
+		Mutate:   MutateTrafficPermissions,
 	})
 }
 
+func MutateTrafficPermissions(res *pbresource.Resource) error {
+	var tp pbauth.TrafficPermissions
+
+	if err := res.Data.UnmarshalTo(&tp); err != nil {
+		return resource.NewErrDataParse(&tp, err)
+	}
+
+	changed := false
+
+	for _, p := range tp.Permissions {
+		for _, s := range p.Sources {
+			if updated := normalizedTenancyForSource(s, res.Id.Tenancy); updated {
+				changed = true
+			}
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	return res.Data.MarshalFrom(&tp)
+}
+
+func normalizedTenancyForSource(src *pbauth.Source, parentTenancy *pbresource.Tenancy) bool {
+	var changed bool
+
+	if t, c := defaultedSourceTenancy(src, parentTenancy); c {
+		src.Partition = t.Partition
+		src.Peer = t.PeerName
+		changed = true
+	}
+
+	for _, e := range src.Exclude {
+		if t, c := defaultedSourceTenancy(e, parentTenancy); c {
+			e.Partition = t.Partition
+			e.Peer = t.PeerName
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+func defaultedSourceTenancy(s pbauth.SourceToSpiffe, parentTenancy *pbresource.Tenancy) (*pbresource.Tenancy, bool) {
+	if !isLocalPeer(s.GetPeer()) {
+		return nil, false
+	}
+
+	tenancy := pbauth.SourceToTenancy(s)
+	origTenancy := proto.Clone(tenancy).(*pbresource.Tenancy)
+	// This uses partition tenancy and traffic permissions use namespace tenancy. This is because namespace  can be empty.
+	resource.DefaultTenancy(tenancy, parentTenancy, resource.DefaultPartitionedTenancy())
```

Comment:
`defaultTenancy` was formerly an unexported method, and its only caller had done a `proto.Clone` on the `parentTenancy` because it _may_ mutate it if the parent was not already defaulted.

You may want to clone that one too.